File: July (Henry LIn)/unit_test/calculator/calculator.py
# text based calculator by python (+,-,*,/)
class Calculator:
    def __init__(self, expression):
        self.tokens = expression.replace(" ", "") # remove space in expression
        self.index = 0 # track the current token
     
    def evaluate(self): 
       return self.expression()
    
    def expression(self): # identify the addition and subtraction
        result = self.term() 
        while self.index < len(self.tokens) and self.tokens[self.index] in ('+', '-'):
            if self.tokens[self.index] == '+':
                self.index += 1
                result += self.term()
            elif self.tokens[self.index] == '-':
                self.index += 1
                result -= self.term()
        return result
    
    def term(self): # identify the multiplication and division
        result = self.factor()
        while self.index < len(self.tokens) and self.tokens[self.index] in ('*', '/'):
            if self.tokens[self.index] == '*':
                self.index += 1
                result *= self.factor()
            elif self.tokens[self.index] == '/':
                self.index += 1
                result /= self.factor()
        return result

    def factor(self): # identify the number or expression in parentheses
        if self.tokens[self.index] == '(':
            self.index += 1
            result = self.expression()
            self.index += 1
            return result
        else:
            start = self.index
            while self.index < len(self.tokens) and self.tokens[self.index].isdigit():
                self.index += 1
            return int(self.tokens[start:self.index])
        
if __name__ == "__main__":
    expression = "3 + 5 * ( 2 - 8 )"
    calculator = Calculator(expression)
    result = calculator.evaluate()  # Call the evaluate method
    print(result)  # Print the result

File: July (Henry LIn)/calculator (python, dart)/calculator_dart/lib/calculator.dart
int calculate() {
  return 6 * 7;
}

class Calculator {
  String expression;
  int index = 0;
  List<String> tokens;

  Calculator(this.expression) : tokens = expression.replaceAll(' ', '').split('');

  double evaluate() {
    return _expression();
  }

  double _expression() {
    double result = _term();
    while (index < tokens.length && (tokens[index] == '+' || tokens[index] == '-')) {
      if (tokens[index] == '+') {
        index++;
        result += _term();
      } else if (tokens[index] == '-') {
        index++;
        result -= _term();
      }
    }
    return result;
  }

  double _term() {
    double result = _factor();
    while (index < tokens.length && (tokens[index] == '*' || tokens[index] == '/')) {
      if (tokens[index] == '*') {
        index++;
        result *= _factor();
      } else if (tokens[index] == '/') {
        index++;
        result /= _factor();
      }
    }
    return result;
  }

  double _factor() {
    if (tokens[index] == '(') {
      index++;
      double result = _expression();
      index++; // Skip ')'
      return result;
    } else {
      int start = index;
      while (index < tokens.length && _isDigit(tokens[index])) {
        index++;
      }
      return double.parse(tokens.sublist(start, index).join(''));
    }
  }

  bool _isDigit(String char) {
    return '0'.compareTo(char) <= 0 && '9'.compareTo(char) >= 0;
  }
}

void main() {
  String expression = "3 + 5 * ( 2 - 8 )";
  Calculator calculator = Calculator(expression);
  double result = calculator.evaluate();
  print("Result: $result");
}